src/hotspot/share/runtime/handshake.cpp

```diff
@@ -443,22 +447,49 @@ bool HandshakeState::operation_pending(HandshakeOperation* op) {
   return _queue.contains(mo);
 }
 
-static bool no_suspend_filter(HandshakeOperation* op) {
-  return !op->is_suspend();
+// Filters
+static bool non_self_queue_filter(HandshakeOperation* op) {
+  return !op->is_async();
+}
+static bool no_async_exception_filter(HandshakeOperation* op) {
+  return !op->is_async_exception();
+}
+static bool async_exception_filter(HandshakeOperation* op) {
+  return op->is_async_exception();
+}
+static bool is_ThreadDeath_filter(HandshakeOperation* op) {
+  return op->is_ThreadDeath();
+}
+static bool no_suspend_no_async_exception_filter(HandshakeOperation* op) {
+  return !op->is_suspend() && !op->is_async_exception();
 }
 
-HandshakeOperation* HandshakeState::get_op_for_self(bool allow_suspend) {
+HandshakeOperation* HandshakeState::get_op_for_self(bool allow_suspend, bool check_async_exception) {
```

Comment:
Nice table! Yes, that's all correct.

Reply:
Thanks for confirming.

Reply:
My original thought was to have the filter in the specific handshake:
```
class MyHandshake : public HandshakeClosure {
public:
  MyHandshake(Handle throwable) : HandshakeClosure("MyHandshake") {}
  void do_thread(Thread* thr) {}
  bool executeable_for_context(JavaThreadState from, JavaThreadState to, ...) {
  	return to == _thread_in_Java;
  }
};
```

This would keep what and when together.

Reply:
But what would you do if different Handshakes needed the same filter?

Reply:
I don't see that as a problem Dan. Each handshake operation can have constraints on when that operation can actually be performed. When we reach a particular handshake "poll point" the logic says "give me the next queued handshake operation that is enabled 'here' ". What I hadn't considered was how to express "here" in a way that the handshake operation could use it to decide whether it was enabled or not i.e. whether there current "poll point" allows for throwing an async exception.
Anyway this is future work.

Reply:
Filed 8284017.